canonical/dex-auth-operator
# Convert to sidecar

This PR rewrites the charm from podspec into sidecar and upgrades the dex image to v2.31.2, so that it can support k8s v1.22.
It also attempts to fix [#31](https://github.com/canonical/dex-auth-operator/issues/31), [#45](https://github.com/canonical/dex-auth-operator/issues/45), [#32](https://github.com/canonical/dex-auth-operator/issues/32), [#43](https://github.com/canonical/dex-auth-operator/issues/43) and [#53](https://github.com/canonical/dex-auth-operator/issues/53).
Some helper methods are removed from the charm as we no longer create resources from yaml manifests. Statefulset's replicas are validated in integration test [tests update in progress].
The `test_upgrade_charm.py` should remain disabled till the sidecar version of dex is promoted to `latest/stable` (as it would fail 'upgrading' to an older version, which is not k8s 1.22+ compliant).
Tests performed:
1. Deploy minimal bundle (on microk8s 1.21, with istio 1.5, `dns storage rbac ingress metallb:10.64.140.43-10.64.140.49` enabled) and log in to dashboard - OK
```
bundle: kubernetes
name: kubeflow
applications:
  dex-auth:
    charm: "/home/ubuntu/dex-auth-operator/dex-auth_ubuntu-20.04-amd64.charm"
    resources:
      dex-auth-image: "dexidp/dex:v2.31.2"
    scale: 1
    trust: true
  istio-ingressgateway:          { charm: istio-gateway,           channel: 1.5/stable, scale: 1, trust: true}
  istio-pilot:                   { charm: istio-pilot,             channel: 1.5/stable, scale: 1, options: { default-gateway: "kubeflow-gateway" } }
  kubeflow-dashboard:            { charm: kubeflow-dashboard,      channel: latest/edge, scale: 1 }
  kubeflow-profiles:             { charm: kubeflow-profiles,       channel: latest/edge, scale: 1 }
  oidc-gatekeeper:               { charm: oidc-gatekeeper,         channel: latest/stable, scale: 1 }
relations:
- [dex-auth:oidc-client, oidc-gatekeeper:oidc-client]
- [istio-pilot:ingress, dex-auth:ingress]
- [istio-pilot:ingress, kubeflow-dashboard:ingress]
- [istio-pilot:ingress, oidc-gatekeeper:ingress]
- [istio-pilot:ingress-auth, oidc-gatekeeper:ingress-auth]
- [istio-pilot:istio-pilot, istio-ingressgateway:istio-pilot]
- [kubeflow-profiles, kubeflow-dashboard]
```
Deploy minimal bundle (on microk8s 1.21 and 1.22, with istio latest/edge) and log in to dashboard - NOK - no gateway created by istio (see https://github.com/canonical/istio-operators/issues/94)
Workaround: deploy istio-pilot rev 75 and istio-gateway rev 54 - OK
2. Deploy minimal bundle (on microk8s 1.22 with istio latest/edge, with above revisions) and log in to dashboard - OK
Note that kubeflow-profiles and dashboard latest/edge are updated for k8s v1.22.
3. Check if the dex service was started correctly - OK
```
$ kubectl logs dex-auth-0 -c dex -n kubeflow
2022-06-20T10:08:50.919Z [pebble] HTTP API server listening on ":38813".
2022-06-20T10:08:50.919Z [pebble] Started daemon.
2022-06-20T10:09:08.643Z [pebble] GET /v1/plan?format=yaml 127.102µs 200
2022-06-20T10:09:08.680Z [pebble] POST /v1/layers 307.65µs 200
2022-06-20T10:09:08.688Z [pebble] GET /v1/files?action=read&path=%2Fetc%2Fdex%2Fconfig.docker.yaml 191.334µs 200
2022-06-20T10:09:08.697Z [pebble] POST /v1/files 7.094348ms 200
2022-06-20T10:09:08.712Z [pebble] POST /v1/services 7.527353ms 202
2022-06-20T10:09:08.741Z [pebble] Service "dex" starting: /usr/local/bin/docker-entrypoint dex serve /etc/dex/config.docker.yaml
[...]
2022-06-20T10:09:18.222Z [dex] time="2022-06-20T10:09:18Z" level=info msg="The custom resource devicetokens.dex.coreos.com already available, skipping create"
2022-06-20T10:09:18.222Z [dex] time="2022-06-20T10:09:18Z" level=info msg="config storage: kubernetes"
2022-06-20T10:09:18.222Z [dex] time="2022-06-20T10:09:18Z" level=info msg="config connector: local passwords enabled"
2022-06-20T10:09:18.222Z [dex] time="2022-06-20T10:09:18Z" level=info msg="config skipping approval screen"
2022-06-20T10:09:18.223Z [dex] time="2022-06-20T10:09:18Z" level=info msg="config refresh tokens rotation enabled: true"
2022-06-20T10:09:18.240Z [dex] time="2022-06-20T10:09:18Z" level=info msg="listening (http) on 0.0.0.0:5556"

```
4. Test if the charm gets updated on config changed - OK
`juju config dex-auth static-username=test`
Config-changed hook is fired and correctly updates the config, allowing to log in with the new username.
5. Verify if the following issues were addressed:
- https://github.com/canonical/dex-auth-operator/issues/31
Since the charm was rewritten to sidecar, the configmap is no longer created.
The `authservice-oidc` is parsed correctly, without quotes:
```
$ kubectl exec -it dex-auth-0 -c dex -n kubeflow -- cat /etc/dex/config.docker.yaml
connectors: null
enablePasswordDB: true
issuer: http://10.64.140.43.nip.io/dex
logger:
  format: text
  level: debug
oauth2:
  skipApprovalScreen: true
staticClients:
- id: authservice-oidc
  name: Ambassador Auth OIDC
  redirectURIs:
  - /authservice/oidc/callback
  secret: Q609Q6KW5KGMWIGMZAOGA2PU16BY3H
staticPasswords:
- email: admin
  hash: $2b$12$LD0V9WIp2Tz.k5djzFoy1eSdmCY4O2HNFykf2zLVEiZjENlZyDhQy
  userID: 333bec64-5f88-4610-b85f-4b602ecea1d2
  username: admin
storage:
  config:
    inCluster: true
  type: kubernetes
web:
  http: 0.0.0.0:5556
```

- https://github.com/canonical/dex-auth-operator/issues/45
`juju config dex-auth connectors="this config should not work"`
The charm gets into error state, not being able to start the service, which is an expected result. Once the config is reverted, dex-auth gets updated correctly.

- https://github.com/canonical/dex-auth-operator/issues/32
The service is no longer created from manifest file.

- https://github.com/canonical/dex-auth-operator/issues/43
We do not use the deployment manifest in sidecar anymore, thus `config_hash` is no longer required as we don't need to trigger its rollout.

- https://github.com/canonical/dex-auth-operator/issues/53
The `oidc-client-relation-broken` error is no longer observed in dex.
